server/src/main/java/nl/inl/blacklab/server/requesthandlers/SearchParameters.java

```diff
@@ -233,6 +237,21 @@ public int getInteger(String name) {
         }
     }
 
+    public int getNumberOfDocs() {
+        String filter = getString("filter");
+        if (StringUtils.isBlank(filter)) {
+            return 0;
+        }
+        //Experiment with this regex as well "(docId:\\p{Graph}+)+"
+        Pattern patt = Pattern.compile("([a-f0-9-]+[)|\\s])+");
+        List<String> allResults = new ArrayList<>();
+        for (MatchResult result : patt.matcher(filter).results()
```

Comment:
Adding Regex Evaluation on the query path? Are we trying to make Blacklab _SLOWER_  :D

Reply:
More seriously: If we want to capture components of the filters, we can promote these into top level structured request properties and then render the lucene filter using them in Blacklab directly.

Another thing we can do is capture them on the Ann side (then we can correlate with request id), no?

Seems wasteful to parse what we construct one step earlier...

Reply:
This regex should not impact request time, p99 is in 2k ms range(in our best scenarios), our P50 is 700+ms. On top of that this is what the JVM is good at optimizing heavily used code. This will get and get optimized thoroughly. I am not particularly worried about this.

That being said I can do the manual parsing of the docs if this is a concern to you @emadelwany

Reply:
The other suggestions could also work, there is tradeoff between how much work we want to put into this particular parameter vs the benefit we get from it. In my analysis parsing at request time provided a good cost/benefit. The correlation is something I considered if we want to graph but analyzing later becomes more of a problem...trade offs.

Reply:
+1 to doing this in some structured fashion. The regex doesn't seem handle `agreementNum:(1)` which we add to certain queries. Easy enough to fix, but would be nice to future proof.

Reply:
The regex was only targeting docs as it would be the most likely parameter impacting request time.  Not sure about agreementNum(at least to begin with).

Also I don't I am following the structured idea, if someone can comment on that would be helpful.

Reply:
Ah, it's not that I want additional logging for `agreementNum`, it's that `agreementNum` -- a property of the request we use -- breaks the regex. Here's an example:

```
(base) brendanroof@ip-192-168-0-3 scratch % cat JavaRegex.java
import java.util.ArrayList;
import java.util.List;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class JavaRegex {
    public static int getNumberOfDocs(String filter) {
        Pattern patt = Pattern.compile("([a-f0-9-]+[)|\\s])+");
        List<String> allResults = new ArrayList<>();
        for (MatchResult result : patt.matcher(filter).results()
                .collect(Collectors.toList())) {
           allResults.add(result.group());
        }
        return allResults.size();
    }

    public static void main(String[] args) {
        String input = "agreementNum:(1) AND section:(\"Title\") AND docId:(17e58226-50f9-42a4-a44d-6c85520b2b58 OR 02bd6d2a-7f69-4617-a8ec-bf197e77d748)";
        int count = getNumberOfDocs(input);

        System.out.println("The number of docs in <" + input + "> is " + count + ".");
    }
}
(base) brendanroof@ip-192-168-0-3 scratch % /usr/local/opt/openjdk@11/bin/javac JavaRegex.java
(base) brendanroof@ip-192-168-0-3 scratch % /usr/local/opt/openjdk@11/bin/java JavaRegex
The number of docs in <agreementNum:(1) AND section:("Title") AND docId:(17e58226-50f9-42a4-a44d-6c85520b2b58 OR 02bd6d2a-7f69-4617-a8ec-bf197e77d748)> is 3.
```

Reply:
We can probably find the Java Lucene filter parser and just use it :D

Reply:
If I understand @emadelwany's idea correctly, he's suggesting that instead of passing a pattern string containing doc ids we pass a JSON list of doc ids and construct the pattern on the BL side. Then we could just take the size of the list.

His alternative of just doing the doc counting on the Ann side also makes sense to me and sounds simpler. A middle ground -- if you specifically need the doc counting in BL (to simplify log analysis maybe?) -- would be to do the count in Ann, but pass the count as another header to BL.

Reply:
Anyway, I was  kind of joking about perf, I'm ok with parsing it but parsing via regex has always been a can of worms in my experience if there isn't a very strict grammar, sending the data in structured form, or using a proper lucene filter parser is probably going to be more robust -- that said if we're being scrappy to get this working quickly and get some results then improve later, you can use regex on maybe only the portion you can about (i.e. from the "doc marker till the end of string or beginning of next marker", something like that?) Just to avoid the kind of error brendan mentions above.

Reply:
@brendan-lexion - yes that's what I meant with my 2 suggestions, spot on. I also like the idea of just passing the number of documents as extra request metadata if that's really all we need here.

Reply:
great, adding the number of docs as header seems like a good idea to me. That would avoid messing with the regex on different inputs.